appengine/genetics/js/blocks.js

```diff
@@ -0,0 +1,463 @@
+      var xmlField = goog.dom.createDom('field', null, name);
+      xmlField.setAttribute('name', 'VAR');
+      var xmlBlock = goog.dom.createDom('block', null, xmlField);
+      xmlBlock.setAttribute('type', 'variables_get');
+      option.callback = Blockly.ContextMenu.callbackFactory(this, xmlBlock);
+      options.push(option);
+    }
+  }
+};
+
+/**
+ * Defines the JavaScript generation for mateAnswer.
+ * @type {!Function}
+ */
+Blockly.JavaScript['genetics_mateAnswer'] =
+    goog.partial(Blockly.JavaScript['genetics_generateMouseFunctionJS_'],
+                 'mateAnswer', 'suitor');
+
+Blockly.Blocks['genetics_getSelf'] = {
+  /**
+   * Block for getting the mouse making the decision.
+   * @this Blockly.Block
+   */
+  init: function() {
+    this.jsonInit({
+      "message0": "%1(%2)",
+      "args0": ["getSelf", ""],
+      "output": "Mouse",
+      "colour": Genetics.Blocks.GENETICS_HUE,
+      "tooltip": BlocklyGames.getMsg('Genetics_getSelfTooltip')
+    });
+  }
+};
+
+/**
+ * Defines the JavaScript generation for getSelf.
+ * @param {Blockly.Block} block
+ * @return {!Array.<string|number>}
+ */
+Blockly.JavaScript['genetics_getSelf'] = function(block) {
+  // Generate JavaScript function that returns the mouse object making the
+  // decision.
+  return ['getSelf()', Blockly.JavaScript.ORDER_FUNCTION_CALL];
+};
+
+Blockly.Blocks['genetics_getMice'] = {
+  /**
+   * Block for getting all mice in the cage.
+   * @this Blockly.Block
+   */
+  init: function() {
+    this.jsonInit({
+      "message0": "%1(%2)",
+      "args0": ["getMice", ""],
+      "output": "Array",
+      "colour": Genetics.Blocks.GENETICS_HUE,
+      "tooltip": BlocklyGames.getMsg('Genetics_getMiceTooltip')
+    });
+  }
+};
+
+/**
+ * Defines the JavaScript generation for getMice.
+ * @param {Blockly.Block} block
+ * @return {!Array.<string|number>}
+ */
+Blockly.JavaScript['genetics_getMice'] = function(block) {
+  // Generate JavaScript for getting mice.
+  return ['getMice()', Blockly.JavaScript.ORDER_FUNCTION_CALL];
+};
+
+/**
+ * Block for getting the properties of a mouse.
+ */
+Blockly.Blocks['genetics_getProperties'] = {
+  /**
+   * Initialization of the getProperties block.
+   * @this Blockly.Block
+   */
+  init: function() {
+    this.jsonInit({
+      "message0": "%1.%2",
+
+      "args0": [
+        {
+          "type": "input_value",
+          "name": "MOUSE",
+          "check": "Mouse"
+        },
+        {
+          "type": "field_dropdown",
+          "name": "PROPERTY",
+          "options": [
+            ['size', 'SIZE'],
+            ['aggressiveness', 'AGGRESSIVENESS'],
+            ['startAggressiveness', 'START_AGGRESSIVENESS'],
+            ['fertility', 'FERTILITY'],
+            ['startFertility', 'START_FERTILITY'],
+            ['sex', 'SEX'],
+            ['age', 'AGE'],
+            ['id', 'ID'],
+            ['pickFightOwner', 'PICK_FIGHT'],
```

Comment:
In the game, a simulation is run where mice breed and their children inherit the three function (pickFight, etc.) from the parents (for each function, the child receives one of the two so there are different owners for each function). These properties are important because a mouse function could be written that takes into account the owners for each of these functions or itself. For example a mouse could only say yes to a mate request if at least one of the other mouse's function owners is the same as it's mateAnswerOwner. I've been switching between representing users as strings and numeric ids. Currently they are represented by numeric ids that I assign to each player.